### PR TITLE
Excon exception handling

### PR DESCRIPTION
**Pull request by stjeanp**
This adds more robust exception handling for Excon requests by implementing a new error class that parses the response and outputs the relevant information. The exceptions are able to be rescued by consumers of the module.

I've also made the request exception handling consistent across the module. All requests now raise these exceptions, instead of some raising Fog::Errors::NotFound while others raised Excon::Errors::NotFound. This has smoothed out some edge cases and presents a much more consistent environment for module consumers to use.

I updated all mock request methods that could throw 404s to use this exception framework for more consistent testing.

I've also made, hopefully, the style and lint fixes that hound would object to. And my fork's travis builds all passed.

### CODE REVIEW

**Comment by stjeanp** [comment]
Everything built cleanly here. I've bumped the version to 0.3.0 due to the exception changes. If it's ok with you, can we merge this and release a new gem?

File: lib/fog/compute/ecloud.rb [hunk 1]
```diff
@@ -266,8 +266,97 @@ def default_organization_uri
         end
       end
 
+      # Custom errors specific to our implementation
+      module Errors
```
Comment (plribeiro3000) [inline lib/fog/compute/ecloud.rb]: My only concern here is this module. This class is growing too big and maybe we could extract it to another file and load it using `autoload`. What do you think?

**Comment by stjeanp** [comment]
I thought the same thing, and started out that way, but gave up pretty quickly. Autoload's pretty much broken due to the layout of the module. I based on the other module that I wrote on the layout of the Rackspace, Google, and other providers in the main fog repo and didn't have these kinds of issues. Based on that I'd suggest that it should be laid out like below. Then the exceptions could be put in their own file and autoloaded without all the path insanity that's there right now. Everything in my other module loads with a simple require, no file gymnastics needed.

I can work on cleaning that up, but it's not something I want to undertake lightly because it's going to break git history, and really make a mess of things until it's all right. If you're ok with that, I'll get rolling on it.

lib/
  fog/
    ecloud.rb
    bin/
      ecloud.rb
    ecloud/
      compute.rb
      core.rb
      version.rb
      models/
        compute/
          <models and collections>
      requests/
        compute/
          <requests>

**Comment by plribeiro3000** [comment]
I did not quite get your point. Are you talking about bad paths for the files?

Like a class `Fog::Ecloud::Something` on a file `fog/something/else/ecloud`?

**Comment by stjeanp** [comment]
Bad paths for the files. They're scattered around in such a way that autoloading doesn't work without resorting to playing with the paths in the require statements. If it were laid out similar to my (terrible) diagram, autoloading would just work.

For example, doing it that way, in order to require the version.rb file, all you'd have to do is put

require 'fog/ecloud/version'

in lib/fog/ecloud.rb

Instead of having to do this:

require File.expand_path("../ecloud/version", __FILE__)

**Comment by plribeiro3000** [comment]
Oh no. That is because of another issue.

Take a look at this commit: c1851227f5fc9155f17f61c23d4bae780855f516
And the original issue is: fog/fog#3430

The short term is that this `File.expand_path` is intentional and we have to keep it.

So don´t worry. They are not scattered around.

BTW i believe that all files follow the namespace of the class. The commit i´ve linked should have fixed all bad paths.

**Comment by stjeanp** [comment]
On 6/24/15 9:51 PM, Paulo Henrique Lopes Ribeiro wrote:
>
> Oh no. That is because of another issue.
>
> Take a look at this commit: c185122 
> <https://github.com/fog/fog-ecloud/commit/c1851227f5fc9155f17f61c23d4bae780855f516>
> And the original issue is: fog/fog#3430 
> <https://github.com/fog/fog/issues/3430>
>
> The short term is that this |File.expand_path| is intentional and we 
> have to keep it.
>
> So don´t worry. They are not scattered around.
>

*cringe*

Ok, I understand, but I'm not happy with that. The other module that I 
wrote (which may never see the light of day for other reasons) works 
fine, even on small VMs, because it (and its consumers) don't use 
"require 'fog'". The core.rb file, that everything below it requires, 
itself requires only 'fog/core' 'fog/xml', and 'fog/json'. So I don't 
seem to be running in to those extreme load time issues. Which is 
exactly what I was going to do with this module. I've put a copy of its 
Gemfile.lock below so that you can see what its dependencies are.

PATH
   remote: .
   specs:
     fog-vzcloud (0.1.0)
       fog-core (~> 1.30)
       fog-json (~> 1.0)
       fog-xml (~> 0.1)
       ipaddress (~> 0.8)

GEM
   remote: https://rubygems.org/
   specs:
     builder (3.2.2)
     excon (0.45.3)
     fog-core (1.31.1)
       builder
       excon (~> 0.45)
       formatador (~> 0.2)
       mime-types
       net-scp (~> 1.1)
       net-ssh (>= 2.1.3)
     fog-json (1.0.0)
       multi_json (~> 1.0)
     fog-xml (0.1.2)
       fog-core
       nokogiri (~> 1.5, >= 1.5.11)
     formatador (0.2.5)
     ipaddress (0.8.0)
     json (1.8.2)
     mime-types (2.6.1)
     mini_portile (0.6.2)
     multi_json (1.11.1)
     net-scp (1.2.1)
       net-ssh (>= 2.6.5)
     net-ssh (2.9.2)
     nokogiri (1.6.6.2)
       mini_portile (~> 0.6.0)
     rake (10.4.2)
     rdoc (4.2.0)
       json (~> 1.4)
     redcarpet (3.2.2)
     yard (0.8.7.6)

PLATFORMS
   ruby

DEPENDENCIES
   bundler (~> 1.6)
   fog-vzcloud!
   json
   rake (~> 10.0)
   rdoc (~> 4.2)
   redcarpet (~> 3.2)
   yard (~> 0.8)

BUNDLED WITH
    1.10.4


> BTW i believe that all files follow the namespace of the class. The 
> commit i´ve linked should have fixed all bad paths.
>

It may work, but it's not doing things in a consistent, logical, and 
easy to understand and expand way. I tried and just about rewrote the 
whole bloody thing to do things in a sane way (at which point we, 
internally, would have abandoned the mainline and used that fork 
exclusively). I'll revisit this again today, to get the exceptions 
included from a separate file and if I can't, that rewrite may still be 
an option. The biggest issue is not being able to use require_relative 
due to the need to build for Ruby 1.8.x. If we could just get rid of 
that requirement, we could scrap all of that and just start using 
require_relative, greatly simplifying things.

https://www.ruby-lang.org/en/news/2013/06/30/we-retire-1-8-7/

What are the numbers for fog-ecloud consumers? How many are using 1.8.x? 
Do they justify all of this extra work?

Pat

-- 
Pat in Lewisville       '15 FLTRXS (Bender)      '97 XLH 883 (Rocinante)
BS#140                       K0OOK                   EKIII rides with me
http://www.pat-st-jean.com/                    http://www.bikerscum.org/

**Comment by stjeanp** [comment]
On 6/24/15 9:51 PM, Paulo Henrique Lopes Ribeiro wrote:
>
> Oh no. That is because of another issue.
>
> Take a look at this commit: c185122 
> <https://github.com/fog/fog-ecloud/commit/c1851227f5fc9155f17f61c23d4bae780855f516>
> And the original issue is: fog/fog#3430 
> <https://github.com/fog/fog/issues/3430>
>
> The short term is that this |File.expand_path| is intentional and we 
> have to keep it.
>

More:

http://guides.rubygems.org/patterns/#loading-code

Pat

-- 
Pat in Lewisville       '15 FLTRXS (Bender)      '97 XLH 883 (Rocinante)
BS#140                       K0OOK                   EKIII rides with me
http://www.pat-st-jean.com/                    http://www.bikerscum.org/

**Comment by plribeiro3000** [comment]
Yeah. There are a lot of stuff going on right now. Let me try to give you more context here since i don't recall all the issue numbers. =s

* We can't drop ruby `1.8.7` yet. This has to come from `fog` itself and until we get `fog 2.0.0` out this is not happening. `fog` depends on all provider gems so if each of them has its own ruby version dependency it is going to be a mess.
* That dependency time issue was from within `fog` that loads all providers. This is not to improve the loading time for one particular provider but for all of them at once.
* The file `core` its not necessary because the provider is not inside `fog` anymore. Now that we rely on bundler to load everything we don't need to be able to load just one service at a time (`Compute`, `DNS`) but rather the main file (`fog/ecloud`, `fog/google`) and bundler together with autoload will do all the magic for us. So we will end up moving its content to the main file that bundler loads like we have here already.

What do you think is a sane, consistent, logical, and easy to understand and expand path?
That require is not beautiful but is good to me at this point considering everything. Can you explain me the issue with it please?

**Comment by stjeanp** [comment]
On 6/25/15 7:28 AM, Paulo Henrique Lopes Ribeiro wrote:
>
> Yeah. There are a lot of stuff going on right now. Let me try to give 
> you more context here since i don't recall all the issue numbers. =s
>
>   * We can't drop ruby |1.8.7| yet. This has to come from |fog| itself
>     and until we get |fog 2.0.0| out this is not happening. |fog|
>     depends on all provider gems so if each of them has its own ruby
>     version dependency it is going to be a mess.
>

Ok, that makes sense.
>
>   * That dependency time issue was from within |fog| that loads all
>     providers. This is not to improve the loading time for one
>     particular provider but for all of them at once.
>

It seems to me that consumers should be using "require 'fog/provider'", 
not "require 'fog'". Doing that, I've never run in to a performance 
issue, even with tiny VMs (512 M/1 core). I could be wrong on this, 
probably am, but it seems to me a waste of effort to try to tune away 
what appears to me to be an incorrect use case. It should be addressed 
in documentation, not code.

>   * The file |core| its not necessary because the provider is not
>     inside |fog| anymore. Now that we rely on bundler to load
>     everything we don't need to be able to load just one service at a
>     time (|Compute|, |DNS|) but rather the main file (|fog/ecloud|,
>     |fog/google|) and bundler together with autoload will do all the
>     magic for us. So we will end up moving its content to the main
>     file that bundler loads like we have here already.
>

Yep, that's what I was trying to do. Like I said, loading everything 
every time is wasteful. Load the provider you need, and only the 
provider you need. Once a module is capable of doing that, it should not 
be using these indirections. They're a mess, they're difficult to 
maintain, and they're not best practices. My other module's vzcloud.rb 
file requires core.rb and each of the services.rb files. It's all self 
contained, works without a "require 'fog'" and has no performance issues 
due to loading unneeded modules.

> What do you think is a sane, consistent, logical, and easy to 
> understand and expand path?
>

I don't consider anything that requires path expansion such as what's 
being done here to be sane, consistent, or logical. It's difficult to 
read, hard to follow, and inconsistent.

For example, the collections do this:

   require File.expand_path("../location", __FILE__)

instead of doing this:

   require 'fog/compute/ecloud/models/location'

Yes, the second one is longer, but you know exactly where the file is 
relative to the module. You get rid of a (admittedly small) amount of 
overhead with the File object too. And it's simple, concise, and gives 
you all the information you need without having to do any mental 
gymnastics. Seriously, it doesn't even make sense, there are no 
directories below the models directory, so how can you be below it? It's 
a hack to get around path expansion quirks. And it's confusing.

But then again, in my other module, it's 'fog/vzcloud/compute/...' 
instead of 'fog/compute/vzcloud/...' so maybe that has something to do 
with it.

> That require is not beautiful but is good to me at this point 
> considering everything. Can you explain me the issue with it please?
>

It's ugly. It's not the recommended way of doing this. Rubygems says 
don't do it. It makes the code harder to read and follow. It obfuscates 
the physical and logical layout of the module.

I've got a change coming that should move the errors out to their own 
file. It's still a mess, but it removes it from the ecloud.rb file 
because, you're right, it's getting way too bloated.

I'd love to move the mock data out to a mock_helper class, but this 
ridiculous pathing and class naming makes it way too painful to do. It 
needs to be fixed because it's not sustainable the way it is.

Pat

-- 
Pat in Lewisville       '15 FLTRXS (Bender)      '97 XLH 883 (Rocinante)
BS#140                       K0OOK                   EKIII rides with me
http://www.pat-st-jean.com/                    http://www.bikerscum.org/

**Comment by plribeiro3000** [comment]
I agree with your statements here and we do push forward to make users do `require 'fog/provider'`. Thats exactly why i'm sending PR's all over the place to projects that depend on `fog`.

The problem is that there are a lot of providers that were not extracted yet and thus we can move the documentation because this feature (modular providers) is not done yet. Thats why the pain of keep compatibility this way. I know it sucks but we need to have all providers extract before we start telling everybody that the usage is gonna change. Do it right now is confusing since the user has to know if the provider is out. =s

The require stuff is only temporary. Once we do extract all providers we gotta change this back to the correct and expected value. We just need this as a temporary solution.

**Comment by stjeanp** [comment]
Thanks for the explanation. I'm not happy with the situation, but I understand and agree with the decision that's been made. My unhappiness is primarily because there's going to be a lot of rework to get this module in to a more standard form, which could potentially break a bunch of stuff. And I hate rework.

I guess we'll deal with that when we need to... :)

**Comment by plribeiro3000** [comment]
Yeah. i don't like rework too. But thats something we need to live since `fog` is so big that any tiny mistake break the hell of the world out. =s
since i've started this i've sent PR's to [carrierwave](https://github.com/carrierwaveuploader/carrierwave/pull/1620), [route53](https://github.com/hw-cookbooks/route53/pull/75) and [asset_sync](https://github.com/rumblelabs/asset_sync/pull/299). I believe there are quite others that we don't even know its broke yet. =s

Thank you for all your help here and understanding of the situation. =)

Are we good to go here? If so i will just wait for travis result.

**Comment by stjeanp** [comment]
I think we are. The mock tests are passing, my use cases for the puppet module that uses this are passing, so I can't see anything holding it up except successful builds.

I'm still working on live tests using the spec interface. I'm not sure how well that will integrate with travis, though. It's all dummy data, so literally you can't break a live cloudspace, but it does hit the production api servers, so you get real tests instead of mock data. I'm using the heck out of it for the puppet module testing, which is how I've found some of these issues.

What're your thoughts on integrating that in to the main build tests? I've got one set done, to test authentication. But I'm hesitant to just submit a pull request, because the travis build will make calls out to the public api server.

**Comment by stjeanp** [comment]
I forgot to add that using the live spec interface for tests would help us get better coverage analysis as well, since there are things that can't be done with the mock approach, such as authentication.

**Comment by plribeiro3000** [comment]
Hmm. on [`fog-xenserver`](https://github.com/fog/fog-xenserver) i've used VCR. Not sure if is the best solution out there but @ihmccreery is also looking for the same answer. =)

I don't like the idea of hitting production servers on tests because you will need to provide credentials which won't be safe. Even with Travis variables. The user can still do a puts and send a PR. =s

**Comment by stjeanp** [comment]
On 6/25/15 12:04 PM, Paulo Henrique Lopes Ribeiro wrote:
>
> Hmm. on |fog-xenserver| <https://github.com/fog/fog-xenserver> i've 
> used VCR. Not sure if is the best solution out there but @ihmccreery 
> <https://github.com/ihmccreery> is also looking for the same answer. =)
>
> I don't like the idea of hitting production servers on tests because 
> you will need to provide credentials which won't be safe. Even with 
> Travis variables. The user can still do a puts and send a PR. =s
>

Actually, that's not the case with the eCloud live spec interface. It 
was designed to do exactly this kind of operation.

Basically, the eCloud live spec interface takes one of three sets of API 
Key credentials. If you supply access_key and private_key made up of all 
3s, it's a full access account. If they're all 2s, it's a no access 
account. And if they're all 1s, it's a read-only account.

ecloud_access_key: 33333333333333333333333333333333
ecloud_private_key: 333333333333333333333333333333333333333333333333333333

When you hit the live spec interface, you're using a slightly different 
base URI (/cloudapi/spec instead of /cloudapi/ecloud) and you're using 
credentials that they make publicly available in the documentation. It 
never touches live cloud instances, doesn't use credentials that can 
EVER touch a live cloudspace, and doesn't actually do any work, just 
responds as if the call was processed, including errors when certain 
values are supplied.

https://services.enterprisecloud.terremark.com/cloudapi/explorer

The docs have some info on it:

http://support.theenterprisecloud.com/kb/redirfile.asp?id=856&SID=

Specifically, look at section 5.3

The TL;DR is that using the live spec is perfectly safe because it's 
simply a simulator that responds just like a real cloud account would, 
without having to have one or create/destroy/change resources.

Pat

-- 
Pat in Lewisville       '15 FLTRXS (Bender)      '97 XLH 883 (Rocinante)
BS#140                       K0OOK                   EKIII rides with me
http://www.pat-st-jean.com/                    http://www.bikerscum.org/

**Comment by ihmccreery** [comment]
You can't hit production servers in tests that run on every PR, because of the issue @plribeiro3000 brought up, but you can hit production servers on CI tests.  The issue is running untrusted code, but as long as the codebase you're running is trusted, you're good to go.

**Comment by stjeanp** [comment]
I'm not explaining myself well. There is absolutely ZERO chance of what Pablo brought up happening because it's physically impossible.

The "live spec" interface is a publicly documented interface for doing this exact kind of testing. The ONLY credentials that it accepts are published in the PDF I linked to previously. You literally CAN'T use real cloud account credentials, they're not accepted. You don't even need to have an account to use it. Click the link, check it out. That's exactly what the tests I'm proposing would be doing.

The "live spec" interface was created to do exactly the kind of testing we need to do. It is physically isolated from the real cloudspaces. It's a production simulator that responds exactly like a real production account would, except it uses dummy data and cannot be used to access a real cloudspace. Ever.

**Comment by ihmccreery** [comment]
Even better!

On Thu, Jun 25, 2015 at 10:38 AM, Patrick St. Jean <notifications@github.com

**Comment by stjeanp** [comment]
Does that satisfy both of your concerns? I know the name's misleading...

**Comment by plribeiro3000** [comment]
LOL. All providers should have a live spec. Thats awesome!
